### ISSUE TITLE
Cleanup GitHub Root Dir

### ISSUE THREAD

**Issue by tresf** [LMMS/lmms#2201]
We have a lot of files in the root of our source tree and many of them are no longer used, redundant, or are input files for cmake.

This is a thread to talk about cleaning these up.  Feedback needed.

The first column is Keep (green), Move (check) or Delete (blank)

Keep?                          |       File                       |   Done? | Notes
-----------------------------|-----------------------------|--------------|---------
:white_check_mark: | `.gitattributes` | :white_check_mark: | Keep
:white_check_mark: | `.gitignore` | :white_check_mark: | Keep
                                 | `.gitmodules`  | :+1:  | Was used for Zyn submodule.  Empty now. Safe to remove.
:white_check_mark:  | `.mailmap`  | :+1:  |  ~~Move, if cmake/our code allows it.~~ Keep per https://github.com/LMMS/lmms/pull/2210#commitcomment-12393264
:white_check_mark: | `.travis.yml`  | :white_check_mark: | Keep
:ballot_box_with_check: | `AUTHORS`  | :+1: | Better suited for ~~data~~ docs
:white_check_mark: | `CMakeLists.txt`  | :white_check_mark:  | Keep
:ballot_box_with_check: | `COPYING`  | :+1:   | Rename to LICENSE.txt?
 :white_check_mark:      | `INSTALL`  | :+1:   | Kept and merged w/ Qt5 ~~Can we just link to the wiki?~~
                                 | `INSTALL.Qt5`  |  :+1:  | ~~Can we just link to the wiki?~~ Merged w/ INSTALL
                                 | `README`  |  :+1: | Can we just keep the markdown file?
:white_check_mark: | `README.md`  | :white_check_mark:  | Keep, needed for GitHub.
                                 | `TODO`  |  :+1: | Safe to remove, per #1964
:ballot_box_with_check: | `build_mingw32` `build_mingw64` | :+1:  | Move to ~~data or~~ `cmake/scripts`.  Update travis, Windows wiki.
                                 | `configure`  |  :+1:  | Not used anymore.  Safe to remove.
:ballot_box_with_check: | `lmms.1`  |  :+1: | Move to ~~data or cmake?~~ `docs`.  Assumes source allows.
:ballot_box_with_check: | `lmms.rc.in`  |  :+1: | Move to `data` ~~or cmake?~~ Assumes source allows.
:ballot_box_with_check: | `lmms.spec.in`  |  :+1:  | RedHat/Fedora `.spec` RPM file?  Move to `data` ~~or cmake?~~  Assumes source allows.
:ballot_box_with_check: | `lmmsconfig.h.in`  |  :+1:  | Move to ~~data or cmake?~~ `src`. Assumes source allows.
:ballot_box_with_check: | `lmmsversion.h.in`  |   :+1: | Move to ~~data or cmake?~~ `src`. Assumes source allows.

**Comment by softrabbit** [comment]
A few opinions. 

Keep?                          |       File                       |   Done? | Notes
-----------------------------|-----------------------------|--------------|---------
 :white_check_mark: | `COPYING`  |   | I'm pretty sure COPYING is a GNU recommendation for naming the file containing the GPL. Not that it binds LMMS in any way, but it seems to be pretty standard.
                                 | `INSTALL`, `INSTALL.Qt5`  |   | At the very least, these could be fusioned into one `INSTALL`.
                                 | `README`  |  | Out of date? Qt 4.6 is required now, file says 4.3. I expect other details to be a bit behind as well. And two files is a bit excessive.
:white_check_mark: | `README.md`  | :white_check_mark:  | Maybe add useful info from `README`; e.g. library dependencies are only listed there.
:ballot_box_with_check: | `build_mingw32`, `build_mingw64` |   | These mingw scripts could probably be incorporated in `CMakeLists.txt`.
                                 | `configure`  |   | Has pointed towards `INSTALL`since 2008. Kill it.
:ballot_box_with_check: | `lmms.1`  |  | Move to data or doc, man pages are documentation.

**Comment by tresf** [comment]
Some replies... 

Keep?                          |       File                       |   Done? | Notes
-----------------------------|-----------------------------|--------------|---------
 :white_check_mark: | `COPYING`  |   | [...] COPYING is a GNU recommendation **Yes it appears to be.  Here's a decent article on it.  Seems to be personal preference. http://stackoverflow.com/a/5678716/3196753**
:white_check_mark: | `README.md`  | :white_check_mark:  | Maybe add useful info from `README`; e.g. library dependencies are only listed there.  **Hmm... Do people still use instructions from a text files, or can we safely assume they go to our well-written wiki first?  I agree there is a nice dependency listing, but what good is that in the README?  Shouldn't that be in the installation steps or is my thinking wrong?**
:ballot_box_with_check: | `build_mingw32`, `build_mingw64` |   | These mingw scripts could probably be incorporated in `CMakeLists.txt`.  **Not sure if this is possible per http://www.cmake.org/pipermail/cmake/2008-April/021122.html**

**Comment by Wallacoloo** [comment]
I feel very lowly in neglecting the fancy formatting you've applied, but here's my weigh-in:

`build_mingw32/64` could possibly be moved into data/scripts - at least, that's where we have `create_apple_bundle.sh.in` etc and those seem related.

I think we can safely delete README, as well as both of the INSTALL files. I'm fond of a "don't repeat yourself" approach - all the information is already present in the wiki, where it is also organized better than a simple text file and more easily kept up to date.

I prefer `LICENSE` over `COPYING`, as a license is exactly what the GPL is, and the GPL covers more than just your permissions to copy (e.g. permissions to embed, modify, etc). While you're at it, I would gladly turn a blind eye if you accidentally replaced it with a less horrid, more permissive license.

Beyond moving `lmms.spec.in` elsewhere, I've proposed a PR to remove its "Changelog" section as well: #2203. 

I don't think `lmms.1` should be moved to cmake - that directory looks to be only cmake *code*. I think it would be more appropriate to move this and `lmms.spec.in` somewhere into data. 

A natural place for `lmmsconfig.h.in` and `lmmsversion.h.in` seems like the root of the `src` directory, as these are used to generate header files which are then used exclusively by other files in the src directory.

I can't say much about `lmms.rc.in`, because I can't actually tell what it does. Is this info for the windows installer?

**Comment by curlymorphic** [comment]
Most of my thoughts have already been covered, so no need to repeat them.

Regarding the ReadMe file. I feel this should contain current building instructions, as it is possible for the source code to come from third party origins with no links to our wiki

**Comment by tresf** [comment]
> I can't say much about `lmms.rc.in`, because I can't actually tell what it does. Is this info for the windows installer?

NSIS handles the installer, it appears this information is for [embedding necessary information into the EXE and DLLs](https://github.com/LMMS/lmms/blob/master/src/CMakeLists.txt#L30).  This is mostly used in Windows to fill the "Properties" screen so that when you right click a file, you can get detailed version information.

As all of our `.in` files, CMake processes it first to fill in the necessary fields, then it's passed to the [`windres`](https://sourceware.org/binutils/docs/binutils/windres.html) utility.

Here's where the input file is mentioned:

https://github.com/LMMS/lmms/blob/master/CMakeLists.txt#L371

And our tarball script references it as well here:
https://github.com/LMMS/lmms/blob/master/CMakeLists.txt#L501

**Comment by tresf** [comment]
> I prefer `LICENSE` over `COPYING`, as a license is exactly what the GPL is, and the GPL covers more than just your permissions to copy

Agreed.

> A natural place for `lmmsconfig.h`.in and `lmmsversion.h.in` seems like the root of the `src`

Since the root of `src` is empty, I tend to agree.  BTW, is there a reason why these files can't be merged into one?

> `build_mingw32/64` could possibly be moved into `data/scripts` - at least, that's where we have `create_apple_bundle.sh.in` etc and those seem related.

Yeah, our build toolchains need a home I think.  I'm still not convinced that `cmake` isn't the best place, as they're all tied to `cmake` in one way or another and they aren't `data` in the terms of a resource file that the software needs to run (such as a pixmap).

Since the root of `cmake` is empty, I say we create a new `scripts` folder in there for all building and packaging needs and consider removing it from `data/scripts`.  To me, `data/scripts` would be more relevant for something like a lmms launcher script that is actually deployed to the machine (Such as how Firefox uses a launcher script on Fedora in /usr/bin), but I'd like to hear more opinions. :+1:

**Comment by tresf** [comment]
Please see #2210 (vs. 2201? how confusing!)

What I've done is remove what we've discussed, kept what we've discussed and only changed the source location of the files.  This means cmake will still put them in the same place it has for building, but it will look for them in their new homes.  We should eventually make the build process directory consistent with the source structure, but I'd rather do that in phases if everyone doesn't mind as `#include lmmsversion.h` versus `#include foo/lmmsversion.h` could be much larger impacting and is better suited for a different PR.

I've decided on `cmake/scripts` as the home for our build scripts.  LMK if you disagree.  I will eventually be moving the Apple build scripts there as I feel they're better suited in the cmake area of our structure as they're directly related to building and packaging (even if their reliance on cmake is minimal, it seems better than the data directory).

I'm also optimistic that travis will help iron out the kinks.  As always, feedback appreciated. :+1:

**Comment by tresf** [comment]
Ok, I had missed a few, but the PR should cover them all now.  Please let me know if there are any objections.

**Comment by softrabbit** [comment]
> I've decided on cmake/scripts as the home for our build scripts. LMK if you disagree.

:hand: 

Why not buildtools? It sits there in the root directory, clearly visible with a descriptive name. OK, anything that is run from cmake fits in cmake/scripts as I see it, but at least the mingw_* scripts are pretty much the opposite, they invoke cmake.

**Comment by tresf** [comment]
> at least the mingw_* scripts are pretty much the opposite, they invoke cmake.

I like the `buildtools` home, thanks.  `buildtools/scripts` or simplyi `buildtools`?

**Comment by tresf** [comment]
Quoting a comment from the PR:

> This breaks any git logs for those who have not set their git config to point to the new `.mailmap` location in doc. Let's keep the `.mailmap` file in the root directory.

Duly noted @Lukas-W, thanks.

-Tres

**Comment by Umcaruje** [comment]
I'd vote for simply `buildtools`.

**Comment by tresf** [comment]
I think all concerns have been addressed via https://github.com/tresf/lmms/commit/95f0b29bd0b311007f9d2360b27819be3413f6bd.  Let me know if I missed something.  If not, I'll squash down the commits and merge.

**Comment by tresf** [comment]
@softrabbit 

> Why not buildtools? It sits there in the root directory, clearly visible with a descriptive name. OK, anything that is run from cmake fits in cmake/scripts as I see it, but at least the mingw_* scripts are pretty much the opposite, they invoke cmake.

To this point, should `nsis/FileAssociation.nsh` be removed from the cmake directory?

-Tres

**Comment by tresf** [comment]
@Lukas-W any opinions on a permanent place for the `build_mingw32`, `build_mingw64` scripts?  We have some other scripts -- like the ones we use for apple bundling and packaging and our new and upcoming msys scripts -- that I'd also consider moving to a similar home.

**Comment by Lukas-W** [comment]
`buildtools` sounds fine to me. I'd like `buildscripts` or just `scripts` as well. Note that the (currently) only file in `buildtools`, `bin2res.cpp`,  will be removed when merging #1891 so there's no real point in using that directory just because it's already there.

**Comment by tresf** [comment]
> there's no real point in using that directory just because it's already there.

Right... to me it all seems related to how we build, so cmake seems about as relevant as any, especially considering we have the NSIS script in there.

-Tres

**Comment by Wallacoloo** [comment]
Maybe the folder could be called "buildres" (build resources) so as to be a
suitable location for all of cmake scripts, cmake-populated includes, NSIS
stuff, bash scripts for mingw, and build tools like bin2res, etc.

**Comment by tresf** [comment]
Warning, TL;DR... 

I still like the cmake folder because it is relevant to what the scripts are for, especially considering we currently have our NSIS packaging script in there.

This is what I propose:

 * We switch all unix build scripts to use bash by default using `#!/bin/bash` rather than `#!/bin/sh` and append the`.sh` file suffix for consistency.

```bash
cmake/build_*.sh        # all non-standard build scripts
cmake/install_*.sh      # all non-standard install scripts (Apple needs one of these)
cmake/package_*.sh      # all non-standard package scripts

#    NOTE: 
#
#    Ideally, we'd do everything automatically through CMake
#         e.g. make 
#              make install
#              make package
#              ..etc
#    but some items such as CMake toolchain and Apple packaging currently
#    have good reasoning on remaining external scripts -- at least for now.
```

For example:
```bash
cmake/build_mingw.sh    # builds for win64 on msys2 and mingw.  win32 can be forced via flag
cmake/install_mac.sh    # install target for the current osx platform
cmake/package_mac.sh    # package for the current osx platform
```

But this begs the question, where do the configure files (`.in`) and package specific icon/installer background files reside?  I vote they go into a platform-specific folder via:

```bash
cmake/linux/lmms.desktop        # desktop shortcut for linux installations
cmake/linux/lmms.xml            # file association for linux installations
cmake/linux/lmms.png            # desktop icon for linux installations
cmake/linux/application-x-lmms-project.svg  # project icon for linux installations

cmake/nsis/FileAssociation.nsh  # helper script for nsis (win32, win64) packaging
cmake/nsis/nsis_branding.bmp    # artwork for nsis installer
cmake/nsis/lmms.ico             # desktop icon for nsis installer

cmake/mac/install_mac.sh.in     # install template for the current osx platform
cmake/mac/package_mac.sh.in     # package template for the current osx platform
cmake/mac/lmms.plist.in         # installation instructions template for osx
cmake/mac/project.icns          # project icon for osx
cmake/mac/lmms.icns             # dock icon for osx installations

cmake/msys/msys_helper.sh       # helper script for msys2 environment
cmake/msys/fetch_ppa.sh         # helper script for msys2 environment
cmake/msys/extract_debs.sh      # helper script for msys2 environment
```

**Comment by Wallacoloo** [comment]
@tresf that sounds reasonable.

**Comment by tresf** [comment]
Also, note, if you look at the data directory we have today, it will be much more organized after this as it is currently filled with build-specific files, which I find a bit misleading considering we use the data directory for static resources as well and this would help define what data is actually there for, rather than a miscellaneous catch-all folder.

**Comment by tresf** [comment]
I've cleaned up the data directory and also taken some time to go through the `CONFIGURE` (`.in`) files and force them to use variables instead of hard-coded values... I haven't done source files, just cmake and resource files for now.

For example... 

  * **Before:** `LMMS`
  * **After:**  `${PROJECT_NAME_UCASE}`

I've also switched to spaces for formatting on some vertically aligned areas to fix improper displaying on certain editors.

I'd like to get this merged soon, so any and all feedback is appreciated.
